A29-xds-tls-security.md

```diff
@@ -0,0 +1,411 @@
+The resolver returns a `Cluster` resource for a channel and the CDS for that cluster has the
+[UpstreamTlsContext](https://github.com/envoyproxy/envoy/blob/35702fed462f63a0a237cfbfdf26184272207c11/api/envoy/extensions/transport_sockets/tls/v3/tls.proto#L27)
+containing the security configuration to be applied to all the child policies of that cluster.
+When a channel is using `XdsChannelCredentials`, gRPC processes the `UpstreamTlsContext`
+as described [above](#cds-for-client-side-security).
+
+At this stage the CDS policy uses the flow described above in
+[Use of Bootstrap File by the Certificate Provider Plugin Framework](#use-of-bootstrap-file-by-the-certificate-provider-plugin-framework)
+to fetch the requisite `CertificateProvider`s (one or two as the case may be). These `CertificateProvider`s,
+in one form or another, need to be made available to all the sub-channels of the cluster. How this is
+done is language dependent. For example, C-core uses the channel args to pass the individual
+`CertificateProvider`s whereas Java constructs a dynamic `SslContextProvider` that is directly
+usable by the sub-channel's TLS handshaker to build an `SslContext` for the impending handshake.
+
+##### Server Authorization aka Subject-Alt-Name Checks
+
+If [match_subject_alt_names](https://github.com/envoyproxy/envoy/blob/main/api/envoy/extensions/transport_sockets/tls/v3/common.proto#L373)
+is populated in the
+[combined_validation_context](https://github.com/envoyproxy/envoy/blob/main/api/envoy/extensions/transport_sockets/tls/v3/tls.proto#L251)
+of the received `UpstreamTlsContext` then gRPC validates the SAN entries in the server certificate
+by matching the `match_subject_alt_names` values using the
+[StringMatcher](https://github.com/envoyproxy/envoy/blob/main/api/envoy/type/matcher/string.proto#L20)
+semantics. This is called server authorization because this is how a client "authorizes" a server for
+the connection.
+
+The implementation is language dependent. As an example, Java uses an implementation
+of `X509TrustManager` through the `SslContext` provided to the client sub-channel. The other
+languages use similar hooks provided by the underlying TLS framework to implement server
+authorization.
+
+#### Server Side Flow
+
+The gRPC server xDS flow is described in
+[gRPC Server ](./A27-xds-global-load-balancing.md#grpc-client-architecture). The server
+configuration comes from the
+[`Listener`](https://github.com/envoyproxy/envoy/blob/main/api/envoy/config/listener/v3/listener.proto#L39)
+resource which contains a list of
+[`FilterChain`](https://github.com/envoyproxy/envoy/blob/main/api/envoy/config/listener/v3/listener_components.proto#L193)s.
+The `FilterChain` contains the `DownstreamTlsContext` that is applied to an incoming connection
+after selecting the best-matching `FilterChain` as described in
+[FilterChainMatch](A36-xds-for-servers.md#filterchainmatch). If the server is using `XdsServerCredentials`,
```

Comment:
The server doesn't know if it is using XdsServerCredentials.

Reply:
True. I understand what you are saying and our implementation does incorporate that. But let me know how this should be reworded here.